From 508fd6bbf88981df1e146cb5789e6dd27d1fc0d7 Mon Sep 17 00:00:00 2001
From: Pavel Grunt <pgrunt@redhat.com>
Date: Fri, 10 Jul 2015 10:37:27 +0200
Subject: [PATCH 5/8] Send monitor config if at least one monitor has
 dimensions

If a client (virt-manager, spicy) is not setting display dimensions
and the "resize-guest" property is disabled, spice-gtk sends a wrong
monitor config message where all the monitors have width = heigh = 0
when the agent connects. This message can confuse the guest, in that
case the guest will change the resolution of its monitor.

Regression since 28312b8d1e287a320851e8828825f2ca138d8b0b

Resolves:
https://bugzilla.redhat.com/show_bug.cgi?id=1240721
---
 gtk/channel-main.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/gtk/channel-main.c b/gtk/channel-main.c
index 1ad090f..7ab3407 100644
--- a/gtk/channel-main.c
+++ b/gtk/channel-main.c
@@ -1282,6 +1282,22 @@ static void agent_clipboard_release(SpiceMainChannel *channel, guint selection)
     agent_msg_queue(channel, VD_AGENT_CLIPBOARD_RELEASE, msgsize, msg);
 }
 
+static gboolean any_display_has_dimensions(SpiceMainChannel *channel)
+{
+    SpiceMainChannelPrivate *c;
+    guint i;
+
+    g_return_if_fail(SPICE_IS_MAIN_CHANNEL(channel));
+    c = channel->priv;
+
+    for (i = 0; i < MAX_DISPLAY; i++) {
+        if (c->display[i].width > 0 && c->display[i].height > 0)
+            return TRUE;
+    }
+
+    return FALSE;
+}
+
 /* main context*/
 static gboolean timer_set_display(gpointer data)
 {
@@ -1294,6 +1310,11 @@ static gboolean timer_set_display(gpointer data)
     if (!c->agent_connected)
         return FALSE;
 
+    if (!any_display_has_dimensions(channel)) {
+        SPICE_DEBUG("Not sending monitors config, at least one monitor must have dimensions");
+        return FALSE;
+    }
+
     session = spice_channel_get_session(SPICE_CHANNEL(channel));
 
     /* ensure we have an explicit monitor configuration at least for
-- 
2.5.0

